client-runtime/src/main/java/com/microsoft/rest/v2/http/HttpPipeline.java

```diff
@@ -167,7 +222,90 @@ public Builder withLogger(Logger logger) {
+            return this;
+        }
+
+        /**
+         * Add the provided RequestPolicy factory to this HttpPipeline builder
+         * directly before the first instance of the provided RequestPolicy
+         * factory type. If the provided RequestPolicy factory type is not
+         * found, then the RequestPolicy factory will be added to the end of the
+         * pipeline.
+         * @param requestPolicyFactoryType The RequestPolicy factory type to
+         *                                 search for.
+         * @param requestPolicyFactory The RequestPolicy factory to add.
+         * @return This HttpPipeline builder.
+         */
+        public Builder withRequestPolicyBefore(Class<? extends RequestPolicy.Factory> requestPolicyFactoryType, RequestPolicy.Factory requestPolicyFactory) {
+            int searchIndex = 0;
+            for (final RequestPolicy.Factory factory : requestPolicyFactories) {
+                if (requestPolicyFactoryType.equals(factory.getClass())) {
+                    break;
+                }
+                else {
+                    ++searchIndex;
+                }
+            }
+            final int factoryCount = requestPolicyFactories.size();
+
+            if (searchIndex == factoryCount) {
+                withRequestPolicy(requestPolicyFactory);
+            } else {
+                final int insertIndex = factoryCount - (searchIndex + 1);
+                withRequestPolicy(insertIndex, requestPolicyFactory);
+            }
+
+            return this;
+        }
+
+        /**
+         * Add the provided RequestPolicy factory to this HttpPipeline builder
+         * directly after the first instance of the provided RequestPolicy
+         * factory type. If the provided RequestPolicy factory type is not
+         * found, then the RequestPolicy factory will be added to the end of the
+         * pipeline.
+         * @param requestPolicyFactoryType The RequestPolicy factory type to
+         *                                 search for.
+         * @param requestPolicyFactory The RequestPolicy factory to add.
+         * @return This HttpPipeline builder.
+         */
+        public Builder withRequestPolicyAfter(Class<? extends RequestPolicy.Factory> requestPolicyFactoryType, RequestPolicy.Factory requestPolicyFactory) {
+            int searchIndex = 0;
+            for (final RequestPolicy.Factory factory : requestPolicyFactories) {
+                if (requestPolicyFactoryType.equals(factory.getClass())) {
+                    break;
+                }
+                else {
+                    ++searchIndex;
+                }
+            }
+            final int factoryCount = requestPolicyFactories.size();
+
+            if (searchIndex == factoryCount) {
+                withRequestPolicy(requestPolicyFactory);
+            } else {
+                final int insertIndex = factoryCount - searchIndex;
+                withRequestPolicy(insertIndex, requestPolicyFactory);
```

Comment:
This expands to an insert at (factoryCount - (factoryCount - searchIndex)) which seems a little obtuse. Just insert into requestPolicyFactories.

Reply:
It's a tricky thing to keep straight in general, though, because it's natural to want to keep the factories in "sequential" order from an external POV while reversing them at the appropriate time to build the linked list, etc.

Reply:
Done.